File: tyama_icpc2016dp2G.rb
#!/usr/bin/ruby
M=5000000
MOD=1000000007
inv=[1]*M
fact=[1]*M
factr=[1]*M
2.upto(M-1){|i|
	inv[i]=(MOD-MOD/i)*inv[MOD%i]%MOD
	fact[i]=i*fact[i-1]%MOD
	factr[i]=inv[i]*factr[i-1]%MOD
}
comb=->n,k{
	return 0 if k<0||n<k
	fact[n]*factr[k]*factr[n-k]%MOD
}
h=->n,k{comb[n+k-1,k]}

loop{
	a,b,c,sx,sy=gets.split.map(&:to_i)
	break if a+b+c==0
	if sx<sy || (sx==sy&&a<b)
		a,b=b,a
		sx,sy=sy,sx
	end
	d=sx-sy
	r=0
	if a==0 # xチームのほうが多く得点しているので、xチームが勝ち0ならばyチームも勝ち0でなければ不合理
		if b>0 || sx!=sy
			r=0
		else
			r=h[c,sx]
		end
	elsif b==0
		r=z=h[a,d-a]*h[a+b+c,sx-d]
	else
		([a,d+b].max..sx).each{|w|
			r=(r+h[a,w-a]*h[b,w-d-b]*h[a+b+c,sx-w])%MOD
		}
	end
	p r*comb[a+b+c,a]*comb[b+c,b]%MOD
}

File: tyama_icpc2016dp2G.c
#include <stdio.h>
#define M 5000000
#define MOD 1000000007
int inv[M],fact[M],factr[M];
int comb(int n,int k){
	if(k<0||n<k)return 0;
	return (long long)fact[n]*factr[k]%MOD*factr[n-k]%MOD;
}
int h(int n,int k){return comb(n+k-1,k);}
int main(){
	int a,b,c,sx,sy,i,d,w,r;
	for(i=0;i<M;i++)inv[i]=fact[i]=factr[i]=1;
	for(i=2;i<M;i++){
		inv[i]=(long long)(MOD-MOD/i)*inv[MOD%i]%MOD;
		fact[i]=(long long)fact[i-1]*i%MOD;
		factr[i]=(long long)inv[i]*factr[i-1]%MOD;
	}
	for(;scanf("%d%d%d%d%d",&a,&b,&c,&sx,&sy),a+b+c;){
		if(sx<sy || (sx==sy&&a<b)){
			a^=b,b^=a,a^=b;
			sx^=sy,sy^=sx,sx^=sy;
		}
		d=sx-sy,r=0;
		if(a==0){
			//xチームのほうが多く得点しているので、xチームが勝ち0ならばyチームも勝ち0でなければ不合理
			r=(b>0||sx!=sy) ? 0 : h(c,sx);
		}else if(b==0){
			r=h(a,d-a)*h(a+b+c,sx-d);
		}else{
			for(w=a>d+b?a:d+b;w<=sx;w++){
				r=((long long)h(a,w-a)*h(b,w-d-b)%MOD*h(a+b+c,sx-w)+r)%MOD;
			}
		}
		printf("%d\n",(int)((long long)r*comb(a+b+c,a)%MOD*comb(b+c,b)%MOD));
	}
	return 0;
}